libraries/AP_InertialSensor/AP_InertialSensor_ASM330.cpp

```diff
@@ -0,0 +1,531 @@
+/*
+ *  This program is free software
+*/
+/*
+ *  driver for ASM330 IMUs
+ *    Supported:
+ *      ASM330LHH
+*/
+
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_InertialSensor_ASM330.h"
+#include "AP_InertialSensor_ASM330_registers.h"
+
+#include <utility>
+
+#include <AP_HAL/GPIO.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define GYRO_SCALE      (G_SCALE_2000DPS)
+#define ACCEL_SCALE     (A_SCALE_16G)
+
+AP_InertialSensor_ASM330::AP_InertialSensor_ASM330(AP_InertialSensor &imu,
+                                                   AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                                   enum Rotation rotation)
+    : AP_InertialSensor_Backend(imu)
+    , _dev(std::move(dev))
+    , _rotation(rotation)
+    , _temp_filter(1000, 1)
+{
+    _temp_degc = 0.0F;
+}
+
+AP_InertialSensor_Backend *AP_InertialSensor_ASM330::probe(AP_InertialSensor &_imu,
+                                                           AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                                           enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_InertialSensor_ASM330 *sensor =
+        NEW_NOTHROW AP_InertialSensor_ASM330(_imu,std::move(dev),
+                                                rotation);
+    if (!sensor || !sensor->_init_sensor()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+bool AP_InertialSensor_ASM330::_init_sensor()
+{
+    _spi_sem = _dev->get_semaphore();
+
+    bool success = _hardware_init();
+
+#if ASM330_DEBUG
+    _dump_registers();
+#endif
+    return success;
+}
+
+bool AP_InertialSensor_ASM330::_hardware_init()
+{
+    _spi_sem->take_blocking();
+
+    uint8_t tries;
+    uint8_t whoami;
+
+    // set flag for reading registers
+    _dev->set_read_flag(0x80);
+
+    whoami = _register_read(ASM330_REG_WHO_AM_I);
+    if (whoami != WHO_AM_I) {
+        DEV_PRINTF("ASM330: unexpected acc/gyro WHOAMI 0x%x\n", whoami);
+        goto fail_whoami;
```

Comment:
```suggestion
        return false;
```

goto not required when using WITH_SEMAPHORE